Add Registry to batch transform callback

elephantmipt opened the pull request:
## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [ ] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [ ] Did you run [colab minimal CI/CD](https://colab.research.google.com/drive/1JCGTVvWlrIsLXMPRRRSWiAstSLic4nbA) with `latest` and `minimal` requirements?

<!-- For CHANGELOG separate each item in unreleased section by blank line to reduce collisions -->


## Description

<!-- Add a more detailed description of the changes if needed. -->


## Related Issue

<!-- If your PR refers to a related issue, link it here. -->


## Type of Change

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)


## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

<!-- Thank you for your contribution! -->


### FAQ
Please review the FAQ before submitting an issue:
- [ ] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [ ] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [ ] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [ ] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [ ] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

elephantmipt committed:
add registry

elephantmipt committed:
fix docs

elephantmipt committed:
fix docs and codestyle. Add example

elephantmipt committed:
fix line length :)

elephantmipt committed:
add registry.py

elephantmipt committed:
codestyle

elephantmipt committed:
fix

Scitator commented on catalyst/registry.py:
```diff
@@ -177,7 +177,7 @@ def _torch_functional_loader(r: registry.Registry):
 
     for func in functions:
-        r.add(getattr(F, func), func)
+        r[func] = getattr(F, func)
```
could you please check here?
@bagxi FYI :)

Scitator commented on catalyst/registry.py:
```diff
@@ -144,4 +144,43 @@ def _loggers_loader(r: registry.Registry):
+    import torch.nn.functional as F
+
+    for func in functions:
+        r[func] = getattr(F, func)
```
could not we just add `F.__all__`? :)

Scitator replied:
@elephantmipt let's use something like that - https://github.com/catalyst-team/catalyst/blob/master/catalyst/registry.py#L23#L24
what do you think?

Scitator committed:
Fix data: imports (#1211)

* notebooks update v1

* notebooks update v2

* notebooks update v2

* notebooks update v2

* minimal RL

* minimal RL

* minimal RL 2

* rl update

* rl update

* demo update

* demo update

* demo update

* demo update

* tests update

* tests update

* import fix

* readme

MrNightSky committed:
Change layerwise to layerwise_params (#1210)

* Change layerwise to layerwise_params (rus version)

At the old version was incorrect key layerwise to set lr, weight_decay and other params for different model layers.
Also add example how to add extra params to DataLoader

* Change layerwise to layerwise_params (eng version)

At the old version was incorrect key layerwise to set lr, weight_decay and other params for different model layers.
Also add example how to add extra params to DataLoader

ditwoo committed:
[WIP] faq docs (#1202)

* initial

* docs: engines initial

* docs: engines initial

* fix: python api; feature: config api example

* docs: reverted back

* docs: spelling; engine functions explained

* fix: note in changelog about faq docs

MrNightSky committed:
Update CHANGELOG.md (#1216)

bagxi committed:
fix: to_numpy wrapper for `AdditiveValueMetric` added (#1214)

Scitator committed:
tests update + readme (#1215)

* tests update + readme

* readme

* workflows changed

* workflows

* readme

Dokholyan committed:
Fix compute method (#1206)

* add DynamicBalanceClassSampler

* add DynamicBalanceClassSampler: add usage example

* add DynamicBalanceClassSampler: add tests

* Update catalyst/data/tests/test_sampler.py

* Update catalyst/data/tests/test_sampler.py

* add DynamicBalanceClassSampler: debag tests

* update sampler: add mode

* add example notebook

* sampler: fixes

* samler: docs

* DynamicBalanceClassSampler: fixes

* change import order

* change import order

* Fix compute problem

* Delete sampler

* Replace compute method

* Update segmentation tests

* Codestyle

* Add test for compute

* Add examples

Co-authored-by: Sergey Kolesnikov <scitator@gmail.com>

riohib committed:
Update ddt.rst (#1217)

Change in Case-3 code, Now running the code under name==__main__ (due to python's multiprocessing logic)

Scitator committed:
docker update (#1218)

* docker update

* docker update

Scitator committed:
docs update (#1219)

Commit:
Updated NeptuneLogger docstrings (#1223)

* fixed import in example, changed .ml -> .ai

* fixed long lines, added NeptuneLogger Callback

* fixed NeptuneLogger imports

* fixed newline formating

* fixed line length and import order

* fixed bad quotes

* fixed bad quotes

* vanilla impl [WIP]

* fixed hparams

* applied new structure to run

* corrected log image path

* added log_artifact to catalyst

* Updated log_artifact docstring

* simplified logger, added log artifact

* log to 'experiment' scope is scope is None

* fixed artifact logging

* better structure of the run

* code style adjustments

* added neptune requirements to the workflow

* typo fix

* dropped repetition in requirements

* codestyle fixes

* filename fix

* Corrected docstring to be 1-liner.

* Update catalyst/loggers/neptune.py

Co-authored-by: Sergey Kolesnikov <scitator@gmail.com>

* Final touches & test fixes

* Checkstyle fixes

* Missed code style

* Revert stop to wait

* Fix failing test

* Comment indentation

* fix tests

* Fix

* Fix

* fix

* Fix

* Updated CHANGELOG and loggers docs

* Fix

* Update setup.py

Co-authored-by: Sergey Kolesnikov <scitator@gmail.com>

* updated neptune docstrings

* style fixed

* updated changelog

* typo fix

Co-authored-by: Jakub Czakon <czakon.jakub@gmail.com>
Co-authored-by: Sergey Kolesnikov <scitator@gmail.com>
Co-authored-by: Hubert Jaworski <hubert.jaworski@neptune.ai>
Co-authored-by: Marcin Mycek <herudaio@gmail.com>

elephantmipt committed:
add partial and kwargs

elephantmipt committed:
Merge branch 'master' into registry_to_batch_transform

elephantmipt committed:
remove unnecessary thing

Scitator commented on catalyst/callbacks/batch_transform.py:
```diff
@@ -165,35 +168,50 @@ def __len__(self):
+        if transform_kwargs is not None:
+            transform = partial(transform, **transform_kwargs)
+        if isinstance(transform, str):
+            transform = REGISTRY.get(transform)
```
looks like we need to swap these two ifs :)

Scitator replied:
```suggestion
        if isinstance(transform, str):
            transform = REGISTRY.get(transform)
        if transform_kwargs is not None:
            transform = partial(transform, **transform_kwargs)
```

Scitator commented on catalyst/callbacks/batch_transform.py:
```diff
@@ -219,6 +237,7 @@ def __init__(
             raise TypeError('Expected scope to be on of the ["on_batch_end", "on_batch_start"]')
         self.input_key = input_key
         self.output_key = output_key
+
```
del?

Scitator replied:
```suggestion
```

Scitator commented on tests/pipelines/test_multilabel_classification.py:
```diff
@@ -34,10 +34,7 @@ def train_experiment(device, engine=None):
-                scope="on_batch_end",
-                input_key="logits",
-                output_key="scores",
+                transform="sigmoid", scope="on_batch_end", input_key="logits", output_key="scores",
```
F.sigmoid?

Scitator replied:
```suggestion
                transform="F.sigmoid", scope="on_batch_end", input_key="logits", output_key="scores",
```

Scitator commented on catalyst/callbacks/batch_transform.py:
```diff
@@ -64,8 +67,8 @@ class BatchTransformCallback(Callback):
-                    dl.LambdaPreprocessCallback(
-                        input_key="logits", output_key="scores", transform=torch.sigmoid
+                    dl.BatchTransformCallback(
+                        input_key="logits", output_key="scores", transform="sigmoid",
```
F.sigmoid?

Scitator replied:
```suggestion
                        input_key="logits", output_key="scores", transform="F.sigmoid",
```

Scitator committed:
Update catalyst/callbacks/batch_transform.py

Scitator committed:
Apply suggestions from code review

Scitator commented on catalyst/registry.py:
```diff
@@ -144,4 +144,13 @@ def _loggers_loader(r: registry.Registry):
+def _torch_functional_loader(r: registry.Registry):
+    import torch.nn.functional as F
+
+    r.add_from_module(F, "F")
```
```suggestion
    r.add_from_module(F, ["F."])
```

Scitator commented on README.md:
```diff
@@ -424,7 +424,7 @@ runner.train(
     callbacks=[
         dl.BatchTransformCallback(
-            transform=torch.sigmoid,
+            transform="sigmoid",
```
```suggestion
            transform=torch.sigmoid,
```

Scitator committed:
Apply suggestions from code review

Scitator committed:
Merge branch 'master' into registry_to_batch_transform
